File: src/org/softpres/life/Grid.scala
package org.softpres.life

import java.awt.{Color, Graphics2D}

/**
 * Scala GOL grid, post optimisation.
 * todo boundary grid
 *
 * MacBook:
 * 22300
 * 24300 - no change, afaik
 * 27000 - removed tuples for changes
 *
 * @author kieron
 */
class Grid(val dimX: Int, val dimY: Int) {

  val cellCount = dimX * dimY
  private var grid = new Array[Boolean](cellCount)
  private var dirty = new Dirty
  private var changes = new Changes
  private var painting: Array[Int] = Array()


  final def tick() {
    step()
    commit()
  }

  final def commit() {
    val consumed = changes.consume()
    for (encoded <- consumed) {
      val index = changes.decodeIndex(encoded)
      val alive = changes.decodeAlive(encoded)
      grid(index) = alive
    }
    painting = consumed
  }

  final def cell(x: Int, y: Int): Boolean = {
    if (x <= 0 || x > dimX || y <= 0 || y > dimY) {
      return false
    }
    return grid(index(x, y))
  }

  final def prime(x: Int, y: Int, alive: Boolean) {
    changes += (index(x, y), alive)
    mark(x, y)
  }

  final def activate(x: Int, y: Int) {
    prime(x, y, true)
  }


  private def x(i: Int) = (i % dimX) + 1
  private def y(i: Int) = (i / dimX) + 1

  private def index(x: Int, y: Int) = (y - 1) * dimX + (x - 1)

  private def mark(x: Int, y: Int) {
    val minX = if (x == 1) 0 else -1
    val minY = if (y == 1) 0 else -1
    val maxX = if (x == dimX) 0 else 1
    val maxY = if (y == dimY) 0 else 1

    var xd = minX
    while (xd <= maxX) {
      var yd = minY
      while (yd <= maxY) {
        dirty += index(x+xd, y+yd)
        yd += 1
      }
      xd += 1
    }
  }

  private def step() {
    val current = dirty.consume()

    for (index <- current) {
      val xp = this.x(index)
      val yp = this.y(index)
      val before = cell(xp, yp)
      val n = neighbours(xp, yp)
      val after = alive(before, n)
      if (before != after) {
        prime(xp, yp, after)
      }
    }
  }

  private def neighbours(x: Int, y: Int): Int = {
    return neighbour(x, y, -1, -1) +
           neighbour(x, y, -1,  0) +
           neighbour(x, y, -1,  1) +
           neighbour(x, y,  0, -1) +
           neighbour(x, y,  0,  1) +
           neighbour(x, y,  1, -1) +
           neighbour(x, y,  1,  0) +
           neighbour(x, y,  1,  1)
  }

  private def neighbour(x: Int, y: Int, xd: Int, yd: Int): Int = {
    return if (cell(x + xd, y + yd)) 1 else 0
  }

  private def alive(alive: Boolean, neighbours: Int): Boolean = {
    if (alive) {
      if (neighbours <  2) return false
      if (neighbours <= 3) return true
    }
    else {
      return neighbours == 3
    }
    return false
  }

  final def draw(g: Graphics2D, scale: Int) {
    var i = 0
    while (i < painting.length) {
      val encoded = painting(i)
      val index = changes.decodeIndex(encoded)
      val alive = changes.decodeAlive(encoded)

      val colour = if (alive) Color.GREEN else Color.DARK_GRAY
      g.setColor(colour)

      val xp = x(index)
      val yp = y(index)
      val sx = scale * (xp - 1)
      val sy = scale * (yp - 1)

      g.fillRect(sx+1, sy+1, scale-1, scale-1)

      i += 1
    }
    painting = Array()
  }

  def count: Int = {
    var sum = 0
    for (cell <- grid) {
      sum += (if (cell) 1 else 0)
    }
    return sum
  }

  override def toString: String = {
    val rows = grid.sliding(dimX, dimX)
    def cell(alive: Boolean) = if (alive) "X" else "O"
    val strs = rows.map(_.map(cell).mkString)
    val res = strs.mkString("\n")
    res
  }

  class Changes {
    private val changes = new Array[Int](cellCount)
    private var size = 0

    def +=(index: Int, alive: Boolean) {
      changes(size) = encode(index, alive)
      size += 1
    }

    def encode(index: Int, alive: Boolean) = if (alive) index+cellCount else index

    def decodeIndex(value: Int) = if (value >= cellCount) value-cellCount else value
    def decodeAlive(value: Int) = value >= cellCount

    def consume(): Array[Int] = {
      val res = changes.take(size)
      size = 0
      res
    }
  }

  class Dirty {
    private var set = new Array[Boolean](cellCount)
    private val changes = new Array[Int](cellCount)
    private var size = 0

    def +=(index: Int) {
      if (!set(index)) {
        set(index) = true
        changes(size) = index
        size += 1
      }
    }

    // todo replace with push/pop by adding "start" position?

    def consume(): Array[Int] = {
      val res = changes.take(size+1) //todo optimise?
      set = new Array[Boolean](cellCount)
      size = 0
      res
    }
  }

}

File: srcjs/grid.js

/**
 * JavaScript GOL grid, post optimisation.
 * todo boundary grid
 *
 * MacBook:
 * ?
 *
 */
function Grid(dimX, dimY) {

  var cellCount = dimX * dimY
  var grid = newArray(cellCount, false)
  var dirty = new FlipSet(cellCount)
  var changes = new Changes
  var trail = new Trail(cellCount)
  var trailDisabled = false
  /** Array of only the changes required to update the grid. */
  var painting = null

  this.testing = function() {
    trail.skip = 1
    trail.optimise = false
  }

  this.clear = function() {
    grid = newArray(cellCount, false)
    dirty.consume()
    changes.consume()
    painting = null
    this.draw()
  }


  this.count = function() {
    var sum = 0
    for (var cell in grid) {
      sum += (cell ? 1 : 0)
    }
    return sum
  }

  this.cell = function(x, y) {
    if (x <= 0 || x > dimX || y <= 0 || y > dimY) {
      return false
    }
    return grid[index(x, y)]
  }

  this.activate = function(x, y) {
    this.prime(x, y, true)
  }

  function x(i) { return (i % dimX) + 1 }
  function y(i) { return Math.floor(i / dimX) + 1 }
  function index(x, y) { return (y - 1) * dimX + (x - 1) }


  this.tick = function(clock) {
    this.step(clock)
    this.commit()
  }

  this.step = function() {
    var current = dirty.consume()

    for (var i = 0; i < current.size(); i++) {
      var index = current.get(i)
      var xp = x(index)
      var yp = y(index)
      var before = this.cell(xp, yp)
      var n = this.neighbours(xp, yp)
      var after = this.alive(before, n)
      if (before !== after) {
        this.prime(xp, yp, after)
        this.updateTrailOnChange(index, after)
      }
    }
  }

  this.neighbours = function(x, y) {
    return this.neighbour(x, y, -1, -1) +
           this.neighbour(x, y, -1,  0) +
           this.neighbour(x, y, -1,  1) +
           this.neighbour(x, y,  0, -1) +
           this.neighbour(x, y,  0,  1) +
           this.neighbour(x, y,  1, -1) +
           this.neighbour(x, y,  1,  0) +
           this.neighbour(x, y,  1,  1)
  }

  this.neighbour = function(x, y, xd, yd) {
    return (this.cell(x + xd, y + yd)) ? 1 : 0
  }

  this.alive = function(alive, neighbours) {
    if (alive) {
      if (neighbours <  2) return false
      if (neighbours <= 3) return true
    }
    else {
      return neighbours === 3
    }
    return false
  }

  this.prime = function(x, y, alive) {
    changes.add(index(x, y), alive)
    mark(x, y)
  }

  var mark = function(x, y) {
    var minX = (x === 1) ? 0 : -1
    var minY = (y === 1) ? 0 : -1
    var maxX = (x === dimX) ? 0 : 1
    var maxY = (y === dimY) ? 0 : 1

    for (var xd = minX; xd <= maxX; xd++) {
      for (var yd = minY; yd <= maxY; yd++) {
        dirty.add(index(x+xd, y+yd))
      }
    }
  }

  /* If a cell has changed, we need to update the trail data. */
  this.updateTrailOnChange = function(index, alive) {
    if (trailDisabled) {
      return
    }

    // Just born, so remove from trail
    if (alive) {
      trail.remove(index)
    }
    // Just died so add or refresh trail
    else {
      trail.refresh(index)
    }
  }

  this.commit = function() {
    var consumed = changes.consume()
    for (var i=0; i<consumed.size(); i++) {
      var encoded = consumed.get(i)
      var index = changes.decodeIndex(encoded)
      var alive = changes.decodeAlive(encoded)
      grid[index] = alive
    }
    painting = consumed
  }


  this.draw = function(g, scale, clock) {
    // When cell dies, currently draws dead colour, so draw cell change first
    this.drawCells(g, scale)
    this.drawTrail(g, scale, clock)
  }

  this.drawCells = function(g, scale) {
    for (var i = 0; i < painting.size(); i++) {
      var encoded = painting.get(i)
      var index = changes.decodeIndex(encoded)
      var alive = changes.decodeAlive(encoded)

      // If it's died, it will have a trail, so do not need to reset
      var drawRequired = trailDisabled || alive

      if (drawRequired) {
        var colour = alive ? "rgb(0, 255, 0)" : "rgb(64, 64, 64)"
        this.drawCell(index, colour, g, scale)
      }
    }
    painting = null
  }

  this.drawTrail = function(g, scale, clock) {
    if (trailDisabled) {
      return
    }

    var drawCell = this.drawCell
    trail.tick(clock, function(index, count) {
      // var colour = "rgb(0, 0, " + count + ")"
      var colour = trail.colours[count]

      drawCell(index, colour, g, scale)
    })
  }

  this.drawCell = function(index, colour, g, scale) {
    var xp = x(index)
    var yp = y(index)
    var sx = scale * (xp - 1)
    var sy = scale * (yp - 1)

    g.fillStyle = colour
    g.fillRect(sx+1, sy+1, scale-1, scale-1)
  }

/*
  function toString {
    var rows = grid.sliding(dimX, dimX)
    def cell(alive) = if (alive) "X" else "O"
    var strs = rows.map(_.map(cell).mkString)
    var res = strs.mkString("\n")
    return res
  }
*/

  /*
   * A flip array with additional functionality to encode and decode indices.
   */
  function Changes() {
    var changes = FlipArray.create(cellCount)

    this.add = function(index, alive) {
      changes.add(encode(index, alive))
    }

    function encode(index, alive) {
      return alive ? index+cellCount : index
    }

    this.decodeIndex = function(value) {
      return (value >= cellCount) ? value-cellCount : value
    }
    this.decodeAlive = function(value) {
      return value >= cellCount
    }

    this.consume = function() {
      changes.flip()
      changes.reset()
      return changes
    }
  }

  function newArray(size, value) {
    var res = new Array(size)
    res.fill(value)
    return res
  }

}
